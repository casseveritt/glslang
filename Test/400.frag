#version 400 core

in vec2 c2D;
flat in int i;
out vec4 outp;
uniform sampler2D arrayedSampler[5];
uniform usampler2DRect samp2dr;
uniform isampler2DArray isamp2DA;

void main()
{
    vec4 v;
    v = texture(arrayedSampler[i], c2D);
    outp.x = gl_ClipDistance[1];

    ivec2 offsets[4];
    const ivec2 constOffsets[4] = ivec2[4](ivec2(1,2), ivec2(3,4), ivec2(15,16), ivec2(-2,0));
    uvec4 uv4 = textureGatherOffsets(samp2dr, c2D, offsets, 2);  // ERROR, offsets not constant
    uv4 = textureGatherOffsets(samp2dr, c2D, constOffsets, 2);
    vec4 v4 = textureGather(arrayedSampler[0], c2D);
    ivec4 iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 3);
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), i);  // ERROR, last argument not const
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 4);  // ERROR, last argument out of range
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(1), 1+2);
    iv4 = textureGatherOffset(isamp2DA, vec3(0.1), ivec2(i));

    vec4 c = gl_FragCoord;
}

layout(location = 4) in vec4 vl; // ERROR, not supported

#ifdef GL_ARB_separate_shader_objects
#extension GL_ARB_separate_shader_objects : enable
#endif

layout(location = 6) in vec4 vl2;

layout(location = 3) uniform vec3 uv3;

layout(location = 5) in vec4 gl_Color;      // ERROR, layout
noperspective in float gl_ClipDistance[4];  // ERROR, can't change qualifier

layout(origin_upper_left, pixel_center_integer) in vec4 gl_FragCoord;  // ERROR, declared after use

uniform sampler2DRectShadow u2drs;

void foo23()
{
    const ivec2[3] offsets = ivec2[3](ivec2(1,2), ivec2(3,4), ivec2(15,16));

    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), ivec2(c2D));     // ERROR, offset not constant
    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), offsets[1]);
    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), offsets[2]);     // ERROR, offset out of range
    textureProjGradOffset(u2drs, outp, vec2(0.0), vec2(0.0), ivec2(-10, 20)); // ERROR, offset out of range
}

patch in vec4 patchIn;              // ERROR
patch out vec4 patchOut;            // ERROR
