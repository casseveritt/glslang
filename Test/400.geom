#version 400 core

void main()
{
    EmitStreamVertex(1);
    EndStreamPrimitive(0);
    EmitVertex();
    EndPrimitive();
    int id = gl_InvocationID;
}

layout(invocations = 4) in outbn { int a; } bn[]; // ERROR, not on a block
layout(max_vertices = 127) out;
layout(invocations = 4) in;

#extension GL_ARB_separate_shader_objects : enable

in gl_PerVertex {      // testing input arrays with a block redeclaration, see 420.geom for without
    vec4 gl_Position;
    layout(std140, location = 3) patch float gl_PointSize; // ERRORs...
} gl_in[];

void foo()
{
    gl_in.length();  // ERROR
    gl_in[1].gl_Position;
}

in vec4 color[];
in vec4 color2[];
in vec4 colorS[3];
in vec4 colorBad[4];

void foo2()
{
    color.length(); // ERROR
    colorS.length();
}

layout(triangles) in;  // give ERROR just for colorBad

in vec4 color[3];
in vec4 color2[3];
in vec4 colorbad2[2];  // ERROR

void foo3()
{
    gl_in.length();
    color.length();
    color2.length();
    colorS.length();
}

layout(location = 4) in vec4 cva[3];
layout(location = 5) in vec4 cvb[3];
layout(location = 2) in mat3 cmc[3];  // ERROR, collision

patch in vec4 patchIn[];            // ERROR
patch out vec4 patchOut;            // ERROR

in float scalar;  // ERROR, no array

layout(max_vertices = 127, invocations = 4) out;      // ERROR
layout(invocations = 4, max_vertices = 127) in;       // ERROR
layout(max_vertices = 127, invocations = 4) uniform;  // 2 ERRORs

in inblockscalar {
    int a;
} inbls;  // ERROR, not an array

in inblocka {
    int a;
} inbla[17];  // ERROR, wrong array size
