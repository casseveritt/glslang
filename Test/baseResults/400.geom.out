400.geom
Warning, version 400 is not yet complete; some version-specific features are present, but many are missing.
ERROR: 0:12: 'invocations' : can only apply to a standalone qualifier 
ERROR: 0:20: 'patch' : not supported in this stage: geometry
ERROR: 0:20: 'gl_PointSize' : cannot add layout to redeclared block member 
ERROR: 0:20: 'gl_PointSize' : cannot add patch to redeclared block member 
ERROR: 0:25: 'length' :  array must first be sized by a redeclaration or layout qualifier
ERROR: 0:36: 'length' :  array must first be sized by a redeclaration or layout qualifier
ERROR: 0:40: 'triangles' : inconsistent input primitive for array size of colorBad
ERROR: 0:44: 'triangles' : inconsistent input primitive for array size of colorbad2
ERROR: 0:56: 'location' : overlapping use of location 4
ERROR: 0:58: 'patch' : not supported in this stage: geometry
ERROR: 0:59: 'patch' : not supported in this stage: geometry
ERROR: 0:61: 'in' : type must be an array: scalar
ERROR: 0:63: 'invocations' : can only apply to 'in' 
ERROR: 0:64: 'max_vertices' : can only apply to 'out' 
ERROR: 0:65: 'max_vertices' : can only apply to 'out' 
ERROR: 0:65: 'invocations' : can only apply to 'in' 
ERROR: 0:67: 'in' : type must be an array: inbls
ERROR: 0:71: 'triangles' : inconsistent input primitive for array size of inbla
ERROR: 18 compilation errors.  No code generated.


invocations = 4
max_vertices = 127
input primitive = triangles
output primitive = none
ERROR: node is still EOpNull!
0:3  Function Definition: main( (void)
0:3    Function Parameters: 
0:5    Sequence
0:5      EmitStreamVertex (void)
0:5        Constant:
0:5          1 (const int)
0:6      EndStreamPrimitive (void)
0:6        Constant:
0:6          0 (const int)
0:7      EmitVertex (void)
0:8      EndPrimitive (void)
0:9      Sequence
0:9        move second child to first child (int)
0:9          'id' (int)
0:9          'gl_InvocationID' (in int)
0:23  Function Definition: foo( (void)
0:23    Function Parameters: 
0:25    Sequence
0:25      Constant:
0:25        1 (const int)
0:26      gl_Position: direct index for structure (in 4-component vector of float)
0:26        direct index (block{in 4-component vector of float gl_Position, in float gl_PointSize})
0:26          'gl_in' (in 3-element array of block{in 4-component vector of float gl_Position, in float gl_PointSize})
0:26          Constant:
0:26            1 (const int)
0:26        Constant:
0:26          0 (const int)
0:34  Function Definition: foo2( (void)
0:34    Function Parameters: 
0:36    Sequence
0:36      Constant:
0:36        1 (const int)
0:37      Constant:
0:37        3 (const int)
0:46  Function Definition: foo3( (void)
0:46    Function Parameters: 
0:48    Sequence
0:48      Constant:
0:48        3 (const int)
0:49      Constant:
0:49        3 (const int)
0:50      Constant:
0:50        3 (const int)
0:51      Constant:
0:51        3 (const int)
0:?   Linker Objects
0:?     'bn' (in 3-element array of block{in int a})
0:?     'gl_in' (in 3-element array of block{in 4-component vector of float gl_Position, in float gl_PointSize})
0:?     'color' (in 3-element array of 4-component vector of float)
0:?     'color2' (in 3-element array of 4-component vector of float)
0:?     'colorS' (in 3-element array of 4-component vector of float)
0:?     'colorBad' (in 4-element array of 4-component vector of float)
0:?     'colorbad2' (in 2-element array of 4-component vector of float)
0:?     'cva' (layout(location=4 ) in 3-element array of 4-component vector of float)
0:?     'cvb' (layout(location=5 ) in 3-element array of 4-component vector of float)
0:?     'cmc' (layout(location=2 ) in 3-element array of 3X3 matrix of float)
0:?     'patchIn' (patch in 3-element array of 4-component vector of float)
0:?     'patchOut' (layout(stream=0 ) patch out 4-component vector of float)
0:?     'scalar' (in float)
0:?     'inbls' (in block{in int a})
0:?     'inbla' (in 17-element array of block{in int a})


Linked geometry stage:

ERROR: Linking geometry stage: At least one shader must specify an output layout primitive

invocations = 4
max_vertices = 127
input primitive = triangles
output primitive = none

